src/update_agent/actor.rs

```diff
@@ -0,0 +1,117 @@
+//! Update agent actor.
+
+use super::{UpdateAgent, UpdateAgentState};
+use actix::prelude::*;
+use failure::Error;
+use futures::future;
+use log::trace;
+
+impl Actor for UpdateAgent {
+    type Context = Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        trace!("update agent started");
+
+        // Kick-start the state machine.
+        Self::tick_now(ctx);
+    }
+}
+
+pub(crate) struct RefreshTick {}
+
+impl Message for RefreshTick {
+    type Result = Result<(), Error>;
+}
+
+impl Handler<RefreshTick> for UpdateAgent {
+    type Result = ResponseActFuture<Self, (), Error>;
+
+    fn handle(&mut self, _msg: RefreshTick, ctx: &mut Self::Context) -> Self::Result {
+        trace!("update agent tick, current state: {:?}", self.state);
+        let prev_state = self.state.clone();
+
+        let state_action = match self.state {
+            UpdateAgentState::StartState => self.initialize(),
+            UpdateAgentState::Initialized => self.try_steady(),
+            UpdateAgentState::Steady => self.try_check_updates(),
+            UpdateAgentState::_EndState => self.finalized(),
```

Comment:
Is the idea down the road that `EndState` happens if one of the state transition hits a failure? Or is it when checking for updates yields an update and it's ready to reboot?

Reply:
If a state transition faults, it stays in the current state, logs an error and re-tries on the next tick.

The EndState is reached (not now, but once the whole state-machine is complete) when the agent manages to get a successful finalization from rpm-ostree. At that point, it will sit in the final state, waiting for reboot.